File: algorithm/cpp/654.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
	TreeNode *buildTree(vector<int>::iterator begin, vector<int>::iterator end) {
		if (begin >= end) return 0;
		auto itr = max_element(begin, end);
		TreeNode *node = new TreeNode(*itr);
		node->left = buildTree(begin, itr);
		node->right = buildTree(itr+1, end);
		return node;
	}
public:
	TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
		return buildTree(nums.begin(), nums.end());
	}
};

File: algorithm/golang/654.go


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(nums []int, begin, end int) *TreeNode {
    if begin >= end {
        return nil
    }
    max_i := begin

    for i := begin+1; i < end; i++ {
        if nums[max_i] < nums[i] {
            max_i = i
        }
    }

    return &TreeNode{
        nums[max_i],
        buildTree(nums, begin, max_i),
        buildTree(nums, max_i+1, end),
    }
}

func constructMaximumBinaryTree(nums []int) *TreeNode {
    return buildTree(nums, 0, len(nums))

}
